Modify Example2.hs to be loadable in current status

nwtgck opened the pull request:
Hi, developers.

I just modified `Example2.hs` to be loadable in current status. This change contains my dirty technique to write parsable quote.

NOTE:  I think you don't have to merge this change. This is just my idea. But if you like it, you can merge this.

## My test command

This is my test command to confirm whether loadable or not

```bash
$ cd llvm-hs-quote/
$ stack build
$ stack ghci
... some messages ...
> :load Example2
... some messages ...
[9 of 9] Compiling Example2         ( Example2.hs, interpreted )
Ok, 9 modules loaded.
```


## Changes

Here is explanation of changes. My changes has three methods.

### Method 1. `getelementptr`
 e.g. Replace `getelementptr inbounds i32, i32* %x, i64 %14` => `getelementptr inbounds i32* %x, i64 %14`

### Method 2. `load`

e.g. Replace `load i32, i32* %8, align 4` => `load i32* %8, align 4`

### Method 3.  Add `call void @llvm.donothing()`

Add `@llvm.donothing()` between 

* `getelementptr`  and `load`
* `mul` and `getelementptr`

I think Method 1 & 2 is the specification of llvm-quote-hs, but Method 3 is not beautiful solution. So I gave some indentation before `call void @llvm.donothing()`.

I hope that it will be helpful for other or future library users and be helpful to fix the parser!

nwtgck committed:
Modify Example2 to be loadable in current status

sdiehl commented:
I don't understand why we should add ` call void @llvm.donothing()` between every `gep` and `load`? If there's a parser bug we should fix that.

nwtgck commented:
> If there's a parser bug we should fix that.

I think this is a parser bug because if I comment it out like`; call void @llvm.donothing()`, I had a compile error related to the parser.

Compile error:

```txt
Example2.hs:13:18: error:
    • Example2.hs:24:6:
    parse error on `='
    • In the quasi-quotation:
        [Q.llmod|
; ModuleID = 'saxpy.ll'

; (from: https://llvm.org/docs/LangRef.html#llvm-donothing-intrinsic)
; (from: https://stackoverflow.com/a/29419659/2885946)
declare void @llvm.donothing()

define void @saxpy(i32* noalias nocapture %x, i32* noalias nocapture %y, i32 %a, i64 %i) {
entry:
  %1 = getelementptr inbounds i32* %x, i64 %i
  ;  call void @llvm.donothing()
```

cocreature commented:
I don’t think removing the type argument for `load` and `getelementptr` matches LLVM’s textual IR here. If you take a look at the docs for [getelementptr](https://llvm.org/docs/LangRef.html#getelementptr-instruction) and [load](https://llvm.org/docs/LangRef.html#load-instruction) they definitely do include that argument. So I think all 3 changes are actually bugs in `llvm-hs-quote` that should be fixed rather than change the example to work around those bugs.

sdiehl commented:
Indeed, when I checked in `Example2.hs` it didn't compile because of parser bugs. Even thought it's valid LLVM.

nwtgck commented:
> I don’t think removing the type argument for load and getelementptr matches LLVM’s textual IR here. 

Constant expressions #21 have to be fixed because I followed the same approach as `load i32* %8` and `getelementptr inbounds i32* %x, i64 %14`. I thought this syntax is smarter than original LLVM IR because original one has redundancy. If the purpose of llvm-hs-quote is that **it parses all valid LLVM IR**, I believe these bugs have to be fixed.

cocreature commented:
Unless there are substantial benefits to deviating from LLVM’s syntax, I’d prefer to stick to it. Small deviations with little benefit (e.g. I don’t think eliminating the redundancy here buys us that much) seem more likely to cause confuse users instead of helping them.

nwtgck commented:
> seem more likely to cause confusion than help users.

I think so too. Actually I confused at the first use.

sdiehl commented:
Yeah, I think this is the basic assumption of all of us involved. We jettisoned that loop syntax after all.

nwtgck committed:
Merge branch 'upstream-master' into Example2-modification

nwtgck committed:
Adapt `getelementptr` and `load` instruction to new syntax

nwtgck commented:
I changed `getelementptr` and `load` back in `Example2.hs`

Diff from original and this PR is <https://github.com/llvm-hs/llvm-hs-quote/pull/23/files>. So all I changed is adding `call void @llvm.donothing()`. When I removed it, I had still a compile error.

nwtgck committed:
Add `nounwind` and `readnone` function attributes to `@llvm.donothing` in Example2.hs

cocreature commented:
Closing in favor of #26 which fixes the parse error.
